File: 55-jump_game.py
# time: O(n^2) - For every element, we are looking at the next nums[i] elements to the  right. nums[i] can be at most n (length of array)
# space: O(1)
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        @lru_cache(maxsize=None)
        def traverse(i):
            if i == len(nums) - 1:
                return True

            for step in range(1, nums[i] + 1):
                if traverse(i + step):
                    return True

            return False

        return traverse(0)

File: 55-jump_game.cs
public class Solution {
    public bool CanJump(int[] nums) {
        Dictionary<int, bool> memo = new();
        return dfs(0);

        bool dfs(int i)
        {
            if (memo.ContainsKey(i))
            {
                return memo[i];
            }

            if (i == nums.Length - 1)
            {
                memo[i] = true;
                return true;
            }

            if (nums[i] == 0)
            {
                memo[i] = false;
                return false;    
            }

            for (int j = 1; j <= nums[i]; j++)
            {
                if (dfs(i + j))
                {
                    memo[i + j] = true;
                    return true;
                }
            }

            memo[i] = false;
            return false;
        }
    }
}
